ledger/simulation/simulator.go

```diff
@@ -0,0 +1,215 @@
+}
+
+// LookupLatest would implicitly use the latest round in the _underlying_
+// Ledger, it would give wrong results if that ledger has moved forward. But it
+// should never be called, as the REST API is the only code using this function,
+// and the REST API should never have access to a simulatorLedger.
+func (l simulatorLedger) LookupLatest(addr basics.Address) (basics.AccountData, basics.Round, basics.MicroAlgos, error) {
+	err := errors.New("unexpected call to LookupLatest")
+	return basics.AccountData{}, 0, basics.MicroAlgos{}, err
+}
+
+// ==============================
+// > Simulator Errors
+// ==============================
+
+// SimulatorError is the base error type for all simulator errors.
+type SimulatorError struct {
+	err error
+}
+
+func (s SimulatorError) Error() string {
+	return s.err.Error()
+}
+
+func (s SimulatorError) Unwrap() error {
+	return s.err
+}
+
+// InvalidTxGroupError occurs when an invalid transaction group was submitted to the simulator.
+type InvalidTxGroupError struct {
+	SimulatorError
+}
+
+// EvalFailureError represents an error that occurred during evaluation.
+type EvalFailureError struct {
+	SimulatorError
+}
+
+// ==============================
+// > Simulator
+// ==============================
+
+// Simulator is a transaction group simulator for the block evaluator.
+type Simulator struct {
+	ledger simulatorLedger
+}
+
+// MakeSimulator creates a new simulator from a ledger.
+func MakeSimulator(ledger *data.Ledger) *Simulator {
+	return &Simulator{
+		ledger: simulatorLedger{ledger, ledger.Latest()},
+	}
+}
+
+func txnHasNoSignature(txn transactions.SignedTxn) bool {
+	return txn.Sig.Blank() && txn.Msig.Blank() && txn.Lsig.Blank()
+}
+
+type missingSigInfo struct {
+	index    int
+	authAddr basics.Address
+}
+
+// A randomly generated private key. The actual value does not matter, as long as this is a valid
+// private key.
+var proxySigner = crypto.PrivateKey{
+	128, 128, 92, 23, 212, 119, 175, 51, 157, 2, 165,
+	215, 137, 37, 82, 42, 52, 227, 54, 41, 243, 67,
+	141, 76, 208, 17, 199, 17, 140, 46, 113, 0, 159,
+	50, 105, 52, 77, 104, 118, 200, 104, 220, 105, 21,
+	147, 162, 191, 236, 115, 201, 197, 128, 8, 91, 224,
+	78, 104, 209, 2, 185, 110, 28, 42, 97,
+}
+
+// check verifies that the transaction is well-formed and has valid or missing signatures.
+// An invalid transaction group error is returned if the transaction is not well-formed or there are invalid signatures.
+// To make things easier, we support submitting unsigned transactions and will respond whether signatures are missing.
+func (s Simulator) check(hdr bookkeeping.BlockHeader, txgroup []transactions.SignedTxn) (bool, error) {
+	proxySignerSecrets, err := crypto.SecretKeyToSignatureSecrets(proxySigner)
+	if err != nil {
+		return false, err
+	}
+
+	// Find and prep any transactions that are missing signatures. We will modify these transactions
+	// to pass signature verification, then restore the original signed transaction before
```

Comment:
Even better, I decided to make a copy of the input array so that the original never gets modified (042704afed1c6de2574a8b9b31b434f4cd74764b). This seems like a cleaner approach.

Reply:
Agreed. Nice choice.